homeassistant/components/tplink/light.py

```diff
@@ -153,52 +161,82 @@ async def async_setup_entry(
+        *,
+        light_module: Light,
+        parent: Device | None = None,
     ) -> None:
-        """Initialize the switch."""
-        super().__init__(device, coordinator)
-        # For backwards compat with pyHS100
-        if device.is_dimmer:
-            # Dimmers used to use the switch format since
-            # pyHS100 treated them as SmartPlug but the old code
-            # created them as lights
-            # https://github.com/home-assistant/core/blob/2021.9.7/homeassistant/components/tplink/common.py#L86
-            self._attr_unique_id = legacy_device_id(device)
-        else:
-            self._attr_unique_id = device.mac.replace(":", "").upper()
+        """Initialize the light."""
+        self._parent = parent
+        super().__init__(device, coordinator, parent=parent)
+        self._light_module = light_module
+        # If _attr_name is None the entity name will be the device name
+        self._attr_name = None if parent is None else device.alias
         modes: set[ColorMode] = {ColorMode.ONOFF}
-        if device.is_variable_color_temp:
+        if light_module.is_variable_color_temp:
             modes.add(ColorMode.COLOR_TEMP)
-            temp_range = device.valid_temperature_range
+            temp_range = light_module.valid_temperature_range
             self._attr_min_color_temp_kelvin = temp_range.min
             self._attr_max_color_temp_kelvin = temp_range.max
-        if device.is_color:
+        if light_module.is_color:
             modes.add(ColorMode.HS)
-        if device.is_dimmable:
+        if light_module.is_dimmable:
             modes.add(ColorMode.BRIGHTNESS)
         self._attr_supported_color_modes = filter_supported_color_modes(modes)
         if len(self._attr_supported_color_modes) == 1:
             # If the light supports only a single color mode, set it now
             self._fixed_color_mode = next(iter(self._attr_supported_color_modes))
-        self._async_update_attrs()
+        self._async_call_update_attrs()
+
+    def _get_unique_id(self) -> str:
+        """Return unique ID for the entity."""
+        # For historical reasons the light platform uses the mac address as
+        # the unique id whereas all other platforms use device_id.
+        device = self._device
+
+        # For backwards compat with pyHS100
+        if device.device_type is DeviceType.Dimmer and isinstance(device, IotDevice):
+            # Dimmers used to use the switch format since
+            # pyHS100 treated them as SmartPlug but the old code
+            # created them as lights
+            # https://github.com/home-assistant/core/blob/2021.9.7/homeassistant/components/tplink/common.py#L86
+            return legacy_device_id(device)
+
+        # Newer devices can have child lights. While there isn't currently
+        # an example of a device with more than one light we use the device_id
+        # for consistency and future proofing
+        if self._parent or device.children:
```

Comment:
Please split the long comments at 80

Reply:
Done in https://github.com/home-assistant/core/pull/120404